{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e29fbd3c-3c8a-4c9b-8cf6-be5ba0907fd3",
   "metadata": {},
   "source": [
    "## Q1. Explain why we have to use the Exception class while creating a Custom Exception."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1f8636e-e96c-467d-9ab8-6b9220953654",
   "metadata": {},
   "source": [
    "Custom exceptions provide you the flexibility to add attributes and methods that are not part of a standard exception. These can store additional information, like an application-specific error code, or provide utility methods that can be used to handle or present the exception to a user. By using the exception class we can use that piece of class again and again."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69dde469-dfde-4f45-8ac9-194f3d0b5061",
   "metadata": {},
   "source": [
    "## Q2. Write a python program to print Python Exception Hierarchy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f2ecf4d0-3f92-4512-b89a-4cfb3a4032ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hierarchy for Built-in exceptions is : \n",
      " BaseException\n",
      "-- Exception\n",
      "---- TypeError\n",
      "------ FloatOperation\n",
      "------ MultipartConversionError\n",
      "---- StopAsyncIteration\n",
      "---- StopIteration\n",
      "---- ImportError\n",
      "------ ModuleNotFoundError\n",
      "------ ZipImportError\n",
      "---- OSError\n",
      "------ ConnectionError\n",
      "-------- BrokenPipeError\n",
      "-------- ConnectionAbortedError\n",
      "-------- ConnectionRefusedError\n",
      "-------- ConnectionResetError\n",
      "---------- RemoteDisconnected\n",
      "------ BlockingIOError\n",
      "------ ChildProcessError\n",
      "------ FileExistsError\n",
      "------ FileNotFoundError\n",
      "------ IsADirectoryError\n",
      "------ NotADirectoryError\n",
      "------ InterruptedError\n",
      "-------- InterruptedSystemCall\n",
      "------ PermissionError\n",
      "------ ProcessLookupError\n",
      "------ TimeoutError\n",
      "------ UnsupportedOperation\n",
      "------ itimer_error\n",
      "------ herror\n",
      "------ gaierror\n",
      "------ SSLError\n",
      "-------- SSLCertVerificationError\n",
      "-------- SSLZeroReturnError\n",
      "-------- SSLWantWriteError\n",
      "-------- SSLWantReadError\n",
      "-------- SSLSyscallError\n",
      "-------- SSLEOFError\n",
      "------ Error\n",
      "-------- SameFileError\n",
      "------ SpecialFileError\n",
      "------ ExecError\n",
      "------ ReadError\n",
      "------ URLError\n",
      "-------- HTTPError\n",
      "-------- ContentTooShortError\n",
      "------ BadGzipFile\n",
      "---- EOFError\n",
      "------ IncompleteReadError\n",
      "---- RuntimeError\n",
      "------ RecursionError\n",
      "------ NotImplementedError\n",
      "-------- ZMQVersionError\n",
      "-------- StdinNotImplementedError\n",
      "------ _DeadlockError\n",
      "------ BrokenBarrierError\n",
      "------ BrokenExecutor\n",
      "-------- BrokenThreadPool\n",
      "------ SendfileNotAvailableError\n",
      "------ ExtractionError\n",
      "------ VariableError\n",
      "---- NameError\n",
      "------ UnboundLocalError\n",
      "---- AttributeError\n",
      "------ FrozenInstanceError\n",
      "---- SyntaxError\n",
      "------ IndentationError\n",
      "-------- TabError\n",
      "---- LookupError\n",
      "------ IndexError\n",
      "------ KeyError\n",
      "-------- NoSuchKernel\n",
      "-------- UnknownBackend\n",
      "------ CodecRegistryError\n",
      "---- ValueError\n",
      "------ UnicodeError\n",
      "-------- UnicodeEncodeError\n",
      "-------- UnicodeDecodeError\n",
      "-------- UnicodeTranslateError\n",
      "------ UnsupportedOperation\n",
      "------ JSONDecodeError\n",
      "------ SSLCertVerificationError\n",
      "------ Error\n",
      "------ UnsupportedDigestmodError\n",
      "------ IllegalMonthError\n",
      "------ IllegalWeekdayError\n",
      "------ ParserError\n",
      "------ ClassNotFound\n",
      "------ ClipboardEmpty\n",
      "------ MessageDefect\n",
      "-------- NoBoundaryInMultipartDefect\n",
      "-------- StartBoundaryNotFoundDefect\n",
      "-------- CloseBoundaryNotFoundDefect\n",
      "-------- FirstHeaderLineIsContinuationDefect\n",
      "-------- MisplacedEnvelopeHeaderDefect\n",
      "-------- MissingHeaderBodySeparatorDefect\n",
      "-------- MultipartInvariantViolationDefect\n",
      "-------- InvalidMultipartContentTransferEncodingDefect\n",
      "-------- UndecodableBytesDefect\n",
      "-------- InvalidBase64PaddingDefect\n",
      "-------- InvalidBase64CharactersDefect\n",
      "-------- InvalidBase64LengthDefect\n",
      "-------- HeaderDefect\n",
      "---------- InvalidHeaderDefect\n",
      "---------- HeaderMissingRequiredValue\n",
      "---------- NonPrintableDefect\n",
      "---------- ObsoleteHeaderDefect\n",
      "---------- NonASCIILocalPartDefect\n",
      "---------- InvalidDateDefect\n",
      "------ MacroToEdit\n",
      "------ InvalidFileException\n",
      "------ UnequalIterablesError\n",
      "------ InvalidVersion\n",
      "------ _InvalidELFFileHeader\n",
      "------ InvalidWheelFilename\n",
      "------ InvalidSdistFilename\n",
      "------ InvalidSpecifier\n",
      "------ InvalidMarker\n",
      "------ UndefinedComparison\n",
      "------ UndefinedEnvironmentName\n",
      "------ InvalidRequirement\n",
      "-------- RequirementParseError\n",
      "------ InvalidVersion\n",
      "---- AssertionError\n",
      "---- ArithmeticError\n",
      "------ FloatingPointError\n",
      "------ OverflowError\n",
      "------ ZeroDivisionError\n",
      "-------- DivisionByZero\n",
      "-------- DivisionUndefined\n",
      "------ DecimalException\n",
      "-------- Clamped\n",
      "-------- Rounded\n",
      "---------- Underflow\n",
      "---------- Overflow\n",
      "-------- Inexact\n",
      "---------- Underflow\n",
      "---------- Overflow\n",
      "-------- Subnormal\n",
      "---------- Underflow\n",
      "-------- DivisionByZero\n",
      "-------- FloatOperation\n",
      "-------- InvalidOperation\n",
      "---------- ConversionSyntax\n",
      "---------- DivisionImpossible\n",
      "---------- DivisionUndefined\n",
      "---------- InvalidContext\n",
      "---- SystemError\n",
      "------ CodecRegistryError\n",
      "---- ReferenceError\n",
      "---- MemoryError\n",
      "---- BufferError\n",
      "---- Warning\n",
      "------ UserWarning\n",
      "-------- GetPassWarning\n",
      "-------- FormatterWarning\n",
      "------ EncodingWarning\n",
      "------ DeprecationWarning\n",
      "-------- ProvisionalWarning\n",
      "------ PendingDeprecationWarning\n",
      "------ SyntaxWarning\n",
      "------ RuntimeWarning\n",
      "-------- ProactorSelectorThreadWarning\n",
      "-------- UnknownTimezoneWarning\n",
      "-------- PEP440Warning\n",
      "------ FutureWarning\n",
      "-------- ProvisionalCompleterWarning\n",
      "------ ImportWarning\n",
      "------ UnicodeWarning\n",
      "------ BytesWarning\n",
      "------ ResourceWarning\n",
      "------ DeprecatedTzFormatWarning\n",
      "------ PkgResourcesDeprecationWarning\n",
      "---- _OptionError\n",
      "---- _Error\n",
      "---- error\n",
      "---- Verbose\n",
      "---- Error\n",
      "---- SubprocessError\n",
      "------ CalledProcessError\n",
      "------ TimeoutExpired\n",
      "---- TokenError\n",
      "---- StopTokenizing\n",
      "---- ClassFoundException\n",
      "---- EndOfBlock\n",
      "---- TraitError\n",
      "---- Error\n",
      "---- Error\n",
      "------ CancelledError\n",
      "------ TimeoutError\n",
      "------ InvalidStateError\n",
      "---- _GiveupOnSendfile\n",
      "---- error\n",
      "---- Incomplete\n",
      "---- TimeoutError\n",
      "---- InvalidStateError\n",
      "---- LimitOverrunError\n",
      "---- QueueEmpty\n",
      "---- QueueFull\n",
      "---- Empty\n",
      "---- Full\n",
      "---- ArgumentError\n",
      "---- ZMQBaseError\n",
      "------ ZMQError\n",
      "-------- ContextTerminated\n",
      "-------- Again\n",
      "-------- InterruptedSystemCall\n",
      "------ ZMQBindError\n",
      "------ NotDone\n",
      "---- PickleError\n",
      "------ PicklingError\n",
      "------ UnpicklingError\n",
      "---- _Stop\n",
      "---- ArgumentError\n",
      "---- ArgumentTypeError\n",
      "---- ConfigError\n",
      "------ ConfigLoaderError\n",
      "-------- ArgumentError\n",
      "------ ConfigFileNotFound\n",
      "---- ConfigurableError\n",
      "------ MultipleInstanceError\n",
      "---- ApplicationError\n",
      "---- error\n",
      "---- TimeoutError\n",
      "---- error\n",
      "---- ReturnValueIgnoredError\n",
      "---- KeyReuseError\n",
      "---- UnknownKeyError\n",
      "---- LeakedCallbackError\n",
      "---- BadYieldError\n",
      "---- ReturnValueIgnoredError\n",
      "---- Return\n",
      "---- InvalidPortNumber\n",
      "---- error\n",
      "---- LZMAError\n",
      "---- RegistryError\n",
      "---- _GiveupOnFastCopy\n",
      "---- Error\n",
      "------ NoSectionError\n",
      "------ DuplicateSectionError\n",
      "------ DuplicateOptionError\n",
      "------ NoOptionError\n",
      "------ InterpolationError\n",
      "-------- InterpolationMissingOptionError\n",
      "-------- InterpolationSyntaxError\n",
      "-------- InterpolationDepthError\n",
      "------ ParsingError\n",
      "-------- MissingSectionHeaderError\n",
      "---- NoIPAddresses\n",
      "---- BadZipFile\n",
      "---- LargeZipFile\n",
      "---- BadEntryPoint\n",
      "---- NoSuchEntryPoint\n",
      "---- DuplicateKernelError\n",
      "---- ErrorDuringImport\n",
      "---- NotOneValueFound\n",
      "---- CannotEval\n",
      "---- OptionError\n",
      "---- BdbQuit\n",
      "---- Restart\n",
      "---- ExceptionPexpect\n",
      "------ EOF\n",
      "------ TIMEOUT\n",
      "---- PtyProcessError\n",
      "---- FindCmdError\n",
      "---- HomeDirError\n",
      "---- ProfileDirError\n",
      "---- IPythonCoreError\n",
      "------ TryNext\n",
      "------ UsageError\n",
      "------ StdinNotImplementedError\n",
      "---- InputRejected\n",
      "---- GetoptError\n",
      "---- ErrorToken\n",
      "---- PrefilterError\n",
      "---- AliasError\n",
      "------ InvalidAliasError\n",
      "---- Error\n",
      "------ InterfaceError\n",
      "------ DatabaseError\n",
      "-------- InternalError\n",
      "-------- OperationalError\n",
      "-------- ProgrammingError\n",
      "-------- IntegrityError\n",
      "-------- DataError\n",
      "-------- NotSupportedError\n",
      "---- Warning\n",
      "---- SpaceInInput\n",
      "---- DOMException\n",
      "------ IndexSizeErr\n",
      "------ DomstringSizeErr\n",
      "------ HierarchyRequestErr\n",
      "------ WrongDocumentErr\n",
      "------ InvalidCharacterErr\n",
      "------ NoDataAllowedErr\n",
      "------ NoModificationAllowedErr\n",
      "------ NotFoundErr\n",
      "------ NotSupportedErr\n",
      "------ InuseAttributeErr\n",
      "------ InvalidStateErr\n",
      "------ SyntaxErr\n",
      "------ InvalidModificationErr\n",
      "------ NamespaceErr\n",
      "------ InvalidAccessErr\n",
      "------ ValidationErr\n",
      "---- ValidationError\n",
      "---- EditReadOnlyBuffer\n",
      "---- _Retry\n",
      "---- InvalidLayoutError\n",
      "---- HeightIsUnknownError\n",
      "---- ParserSyntaxError\n",
      "---- InternalParseError\n",
      "---- _PositionUpdatingFinished\n",
      "---- SimpleGetItemNotFound\n",
      "---- UncaughtAttributeError\n",
      "---- HasNoContext\n",
      "---- ParamIssue\n",
      "---- _JediError\n",
      "------ InternalError\n",
      "------ WrongVersion\n",
      "------ RefactoringError\n",
      "---- OnErrorLeaf\n",
      "---- InvalidPythonEnvironment\n",
      "---- MessageError\n",
      "------ MessageParseError\n",
      "-------- HeaderParseError\n",
      "-------- BoundaryError\n",
      "------ MultipartConversionError\n",
      "------ CharsetError\n",
      "---- Error\n",
      "---- HTTPException\n",
      "------ NotConnected\n",
      "------ InvalidURL\n",
      "------ UnknownProtocol\n",
      "------ UnknownTransferEncoding\n",
      "------ UnimplementedFileMode\n",
      "------ IncompleteRead\n",
      "------ ImproperConnectionState\n",
      "-------- CannotSendRequest\n",
      "-------- CannotSendHeader\n",
      "-------- ResponseNotReady\n",
      "------ BadStatusLine\n",
      "-------- RemoteDisconnected\n",
      "------ LineTooLong\n",
      "---- InteractivelyDefined\n",
      "---- KillEmbedded\n",
      "---- Error\n",
      "------ NoSuchProcess\n",
      "-------- ZombieProcess\n",
      "------ AccessDenied\n",
      "------ TimeoutExpired\n",
      "---- _Ipv6UnsupportedError\n",
      "---- QueueEmpty\n",
      "---- QueueFull\n",
      "---- DebuggerInitializationError\n",
      "---- ExpatError\n",
      "---- Error\n",
      "------ ProtocolError\n",
      "------ ResponseError\n",
      "------ Fault\n",
      "---- ParseBaseException\n",
      "------ ParseException\n",
      "------ ParseFatalException\n",
      "-------- ParseSyntaxException\n",
      "---- RecursiveGrammarException\n",
      "---- ResolutionError\n",
      "------ VersionConflict\n",
      "-------- ContextualVersionConflict\n",
      "------ DistributionNotFound\n",
      "------ UnknownExtra\n",
      "---- _Error\n",
      "---- UnableToResolveVariableException\n",
      "---- InvalidTypeInArgsException\n",
      "-- GeneratorExit\n",
      "-- SystemExit\n",
      "-- KeyboardInterrupt\n",
      "-- CancelledError\n",
      "-- AbortThread\n"
     ]
    }
   ],
   "source": [
    "# import inspect module\n",
    "import inspect\n",
    "\n",
    "# our treeClass function\n",
    "def treeClass(cls, ind = 0):\n",
    "\t\n",
    "\t# print name of the class\n",
    "\tprint ('-' * ind, cls.__name__)\n",
    "\t\n",
    "\t# iterating through subclasses\n",
    "\tfor i in cls.__subclasses__():\n",
    "\t\ttreeClass(i, ind + 2)\n",
    "\n",
    "print(\"Hierarchy for Built-in exceptions is : \")\n",
    "\n",
    "# inspect.getmro() Return a tuple\n",
    "# of class cls’s base classes.\n",
    "\n",
    "# building a tree hierarchy\n",
    "inspect.getclasstree(inspect.getmro(BaseException))\n",
    "\n",
    "# function call\n",
    "treeClass(BaseException)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83f07867-69ee-4710-a6fb-b1fd641e256a",
   "metadata": {},
   "source": [
    "## Q3. What errors are defined in the ArithmeticError class? Explain any two with an example."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9221175f-1a68-4d53-a5d9-38ad007ec5e7",
   "metadata": {},
   "source": [
    "The arithmetic error occurs when an error is encountered during numeric calculations in Python. This includes Zerodivision Error and Floating point error. In addition, zero division error is raised when you divide a numeric value by zero. number = 6 ZeroErro = 6 / 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8ef6649b-e802-48c1-b1b0-e966c7520bc0",
   "metadata": {},
   "outputs": [
    {
     "ename": "ZeroDivisionError",
     "evalue": "integer division or modulo by zero",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m a \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m6\u001b[39m\n\u001b[0;32m----> 2\u001b[0m \u001b[43ma\u001b[49m\u001b[38;5;241;43m/\u001b[39;49m\u001b[38;5;241;43m/\u001b[39;49m\u001b[38;5;241;43m0\u001b[39;49m\n",
      "\u001b[0;31mZeroDivisionError\u001b[0m: integer division or modulo by zero"
     ]
    }
   ],
   "source": [
    "a = 6\n",
    "a//0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f990be1d-f098-47a6-86df-3208957adb37",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5.0\n",
      "25.0\n",
      "15625.0\n",
      "5.960464477539062e+16\n",
      "7.52316384526264e+83\n"
     ]
    },
    {
     "ename": "OverflowError",
     "evalue": "(34, 'Numerical result out of range')",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOverflowError\u001b[0m                             Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[6], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m j \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m5.0\u001b[39m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;241m1\u001b[39m, \u001b[38;5;241m1000\u001b[39m):\n\u001b[0;32m----> 4\u001b[0m     j \u001b[38;5;241m=\u001b[39m \u001b[43mj\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mi\u001b[49m\n\u001b[1;32m      5\u001b[0m     \u001b[38;5;28mprint\u001b[39m(j)\n",
      "\u001b[0;31mOverflowError\u001b[0m: (34, 'Numerical result out of range')"
     ]
    }
   ],
   "source": [
    "j = 5.0\n",
    "\n",
    "for i in range(1, 1000):\n",
    "    j = j**i\n",
    "    print(j)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3fdbcc4-be61-4211-a32c-353baf346351",
   "metadata": {},
   "source": [
    "## Q4. Why LookupError class is used? Explain with an example KeyError and IndexError."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2897cd37-d16d-4e5d-856e-b9532cec6edb",
   "metadata": {},
   "source": [
    "Lookup error is the base class for key error and index error. The key error is raised when a wrong key is used to access a dictionary value, i.e., using a key that is not in the dictionary. This error means that the key you use to access a dictionary is wrong.The index error is raised when you try to access a sequence (index) of a list that does not exist or is out of range.The index error is raised when you try to access a sequence (index) of a list that does not exist or is out of range."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7d33e95d-96e4-42ed-84aa-19b5383bc3f9",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'names'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# KeyError\u001b[39;00m\n\u001b[1;32m      3\u001b[0m a \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mname\u001b[39m\u001b[38;5;124m\"\u001b[39m:\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124marjun\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mage\u001b[39m\u001b[38;5;124m\"\u001b[39m:\u001b[38;5;241m22\u001b[39m}\n\u001b[0;32m----> 4\u001b[0m \u001b[43ma\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mnames\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'names'"
     ]
    }
   ],
   "source": [
    "# KeyError\n",
    "\n",
    "a = {\"name\":\"arjun\",\"age\":22}\n",
    "a[\"names\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "322dcec9-4fe1-464a-ad15-8e22b1728a33",
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m b\u001b[38;5;241m=\u001b[39m[\u001b[38;5;241m1\u001b[39m,\u001b[38;5;241m2\u001b[39m,\u001b[38;5;241m3\u001b[39m,\u001b[38;5;241m4\u001b[39m,\u001b[38;5;241m5\u001b[39m,\u001b[38;5;241m6\u001b[39m]\n\u001b[0;32m----> 2\u001b[0m \u001b[43mb\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m10\u001b[39;49m\u001b[43m]\u001b[49m\n",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "b=[1,2,3,4,5,6]\n",
    "b[10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8d01efd-1454-480c-8a04-a2bb03d67b07",
   "metadata": {},
   "source": [
    "## Q5. Explain ImportError. What is ModuleNotFoundError?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc91b03b-66c8-4838-a303-72e9ab6c216b",
   "metadata": {},
   "source": [
    "Sometimes, Python throws the ModuleNotFoundError afterward. What does this error mean in Python? As the name implies, this error occurs when you're trying to access or use a module that cannot be found. In the case of the title, the \"module named Python\" cannot be found. Python here can be any module."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "238d541a-6c2f-4681-be75-076907a3f624",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'ffdfd'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[10], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mffdfd\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mimport file not found\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'ffdfd'"
     ]
    }
   ],
   "source": [
    "import ffdfd\n",
    "\n",
    "print(\"import file not found\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b374532a-6b0d-4e6c-ae6e-713fe972bcc5",
   "metadata": {},
   "source": [
    "## Q6. List down some best practices for exception handling in python."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82fa50d1-5db7-4dee-9600-cdf299e8407c",
   "metadata": {},
   "source": [
    "1. Use always a specific exception.\n",
    "2. Always print a valid message.\n",
    "3. Alwaws try to log.\n",
    "4. Always avoid to write multiple exception handling.\n",
    "5. Prepare a proper documentation.\n",
    "6. clean-up all the resources."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cdd8cf1-c112-48ab-922f-1abd3d4cc604",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
